File: src/main/java/me/modmuss50/fr/FluxedRedstone.java
package me.modmuss50.fr;

import mcmultipart.multipart.MultipartRegistry;
import me.modmuss50.fr.client.PipeModelBakery;
import me.modmuss50.fr.mutlipart.ItemMultipartPipe;
import me.modmuss50.fr.mutlipart.MultipartPipe;
import me.modmuss50.fr.mutlipart.PipeStateHelper;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.init.Blocks;
import net.minecraft.init.Items;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.FMLCommonHandler;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.registry.GameRegistry;
import net.minecraftforge.fml.relauncher.Side;
import reborncore.RebornCore;
import reborncore.common.util.CraftingHelper;

import java.util.HashMap;

@Mod(modid = "fluxedredstone", name = "FluxedRedstone", version = "@MODVERSION@", dependencies = "required-after:reborncore;required-after:mcmultipart")
public class FluxedRedstone {

    public static HashMap<PipeTypeEnum, Item> itemMultiPipe = new HashMap<>();

    public static PipeStateHelper stateHelper;

    public static FluxedRedstoneCreativeTab creativeTab;

    @Mod.EventHandler
    public void init(FMLInitializationEvent event) {
        stateHelper = new PipeStateHelper();
        creativeTab = new FluxedRedstoneCreativeTab();


        for(PipeTypeEnum typeEnum : PipeTypeEnum.values()){
            MultipartRegistry.registerPart(typeEnum.getClassType(), "fluxedredstone:fluxedPipe." + typeEnum.getFriendlyName());
            itemMultiPipe.put(typeEnum, new ItemMultipartPipe(typeEnum).setCreativeTab(creativeTab).setUnlocalizedName("fluxedredstone.itemFluxedPipe." + typeEnum.getFriendlyName()));
            GameRegistry.registerItem(itemMultiPipe.get(typeEnum), "itemFluxedPipe." + typeEnum.getFriendlyName());
            RebornCore.jsonDestroyer.registerObject(itemMultiPipe.get(typeEnum));
        }

        if(FMLCommonHandler.instance().getSide() == Side.CLIENT){ //Lazy man's proxy
            MinecraftForge.EVENT_BUS.register(new PipeModelBakery());
        }

        CraftingHelper.addShapedOreRecipe(new ItemStack(itemMultiPipe.get(PipeTypeEnum.REDSTONE), 6),
                "GRG", "RSR", "GRG",
                'G', new ItemStack(Blocks.glass_pane),
                'R', new ItemStack(Items.redstone),
                'S', new ItemStack(Blocks.stone));

        CraftingHelper.addShapedOreRecipe(new ItemStack(itemMultiPipe.get(PipeTypeEnum.GOLD), 6),
                "GRG", "RSR", "GRG",
                'G', new ItemStack(Blocks.glass_pane),
                'R', new ItemStack(Items.gold_ingot),
                'S', new ItemStack(itemMultiPipe.get(PipeTypeEnum.REDSTONE)));

        CraftingHelper.addShapedOreRecipe(new ItemStack(itemMultiPipe.get(PipeTypeEnum.BALZE), 4),
                "GRG", "RSR", "GRG",
                'G', new ItemStack(Blocks.iron_bars),
                'R', new ItemStack(Items.blaze_rod),
                'S', new ItemStack(itemMultiPipe.get(PipeTypeEnum.GOLD)));

        CraftingHelper.addShapedOreRecipe(new ItemStack(itemMultiPipe.get(PipeTypeEnum.ENDER), 4),
                "GRG", "RSR", "GRG",
                'G', new ItemStack(Items.gold_ingot),
                'R', new ItemStack(Items.ender_pearl),
                'S', new ItemStack(Items.ghast_tear));


    }

    public static class FluxedRedstoneCreativeTab extends CreativeTabs {

        public FluxedRedstoneCreativeTab() {
            super("fluxedredstone");
        }

        @Override
        public Item getTabIconItem() {
            return itemMultiPipe.get(PipeTypeEnum.ENDER);
        }
    }

}

File: src/main/kotlin/me/modmuss50/fr/FluxedRedstone.kt
package me.modmuss50.fr

import me.modmuss50.fr.client.PipeModelBakery
import me.modmuss50.fr.mutlipart.IC2Interface
import me.modmuss50.fr.mutlipart.ItemMultipartPipe
import me.modmuss50.fr.mutlipart.PipeTypeEnum
import me.modmuss50.fr.mutlipart.TeslaManager
import me.modmuss50.fr.network.FRNetworkHandler
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.init.Blocks
import net.minecraft.init.Items
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.util.ResourceLocation
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.fml.common.FMLCommonHandler
import net.minecraftforge.fml.common.Loader
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.event.FMLInitializationEvent
import net.minecraftforge.fml.common.event.FMLMissingMappingsEvent
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent
import net.minecraftforge.fml.common.registry.GameRegistry
import net.minecraftforge.fml.relauncher.Side
import reborncore.RebornCore
import reborncore.RebornRegistry
import reborncore.common.util.CraftingHelper
import reborncore.mcmultipart.multipart.IMultipart
import reborncore.mcmultipart.multipart.MultipartRegistry
import java.util.*

@Mod(modid = "fluxedredstone", name = "FluxedRedstone", version = "@MODVERSION@", dependencies = "required-after:reborncore;required-after:reborncore-mcmultipart;required-after:forge@[12.18.1.2080,);")
class FluxedRedstone {

    @Mod.EventHandler
    fun preInit(event: FMLPreInitializationEvent) {
        config = Config(event.suggestedConfigurationFile)
        config.preInit(event)
    }

    @Mod.EventHandler
    fun init(event: FMLInitializationEvent) {
        if (!Loader.isModLoaded("Tesla")) {
            teslaSupport = false
        } else if (teslaSupport) {
            teslaManager = TeslaManager()
        }

        if (!Loader.isModLoaded("IC2")) {
            ic2Support = false
        } else {
            ic2Interface = IC2Interface()
            network = FRNetworkHandler()
            network.registerPackets()
        }

        for (typeEnum in PipeTypeEnum.values()) {
            MultipartRegistry.registerPart(typeEnum.getJavaClass() as Class<out IMultipart>?, "fluxedredstone:fluxedPipe." + typeEnum.friendlyName)
            itemMultiPipe.put(typeEnum, ItemMultipartPipe(typeEnum).setCreativeTab(creativeTab).setUnlocalizedName("fluxedredstone.itemFluxedPipe." + typeEnum.friendlyName))
            RebornRegistry.registerItem(itemMultiPipe[typeEnum], ResourceLocation("fluxedredstone", "itemFluxedPipe." + typeEnum.friendlyName))
            RebornCore.jsonDestroyer.registerObject(itemMultiPipe[typeEnum])
        }

        if (FMLCommonHandler.instance().side == Side.CLIENT) { //Lazy man's proxy
            MinecraftForge.EVENT_BUS.register(PipeModelBakery())
        }

        CraftingHelper.addShapedOreRecipe(ItemStack(itemMultiPipe[PipeTypeEnum.REDSTONE], 6),
                "GRG", "RSR", "GRG",
                'G', ItemStack(Blocks.GLASS_PANE),
                'R', ItemStack(Items.REDSTONE),
                'S', ItemStack(Blocks.STONE))

        CraftingHelper.addShapedOreRecipe(ItemStack(itemMultiPipe[PipeTypeEnum.GOLD], 6),
                "GRG", "RSR", "GRG",
                'G', ItemStack(Blocks.GLASS_PANE),
                'R', ItemStack(Items.GOLD_INGOT),
                'S', ItemStack(itemMultiPipe[PipeTypeEnum.REDSTONE]))

        CraftingHelper.addShapedOreRecipe(ItemStack(itemMultiPipe[PipeTypeEnum.BALZE], 4),
                "GRG", "RSR", "GRG",
                'G', ItemStack(Blocks.IRON_BARS),
                'R', ItemStack(Items.BLAZE_ROD),
                'S', ItemStack(itemMultiPipe[PipeTypeEnum.GOLD]))

        CraftingHelper.addShapedOreRecipe(ItemStack(itemMultiPipe[PipeTypeEnum.ENDER], 4),
                "GRG", "RSR", "GRG",
                'G', ItemStack(Items.GOLD_INGOT),
                'R', ItemStack(Items.ENDER_PEARL),
                'S', ItemStack(Items.GHAST_TEAR))
    }

    @Mod.EventHandler
    fun remap(event: FMLMissingMappingsEvent){
        event.all
                .filter { it.type == GameRegistry.Type.ITEM }
                .forEach {
                    for (typeEnum in PipeTypeEnum.values()) {
                        if(it.resourceLocation.resourceDomain == "minecraft" && it.resourceLocation.resourcePath == "itemfluxedpipe." + typeEnum.friendlyName.toLowerCase()){
                            println("Remapping " + it.resourceLocation.resourcePath)
                            it.remap(itemMultiPipe.get(typeEnum))
                        }
                    }
                }
    }


    class FluxedRedstoneCreativeTab : CreativeTabs("fluxedredstone") {

        override fun getTabIconItem(): ItemStack {
            return ItemStack(FluxedRedstone.itemMultiPipe[PipeTypeEnum.ENDER]!!)
        }
    }


    companion object {
        var creativeTab = FluxedRedstoneCreativeTab()

        lateinit var config : Config

        lateinit var teslaManager : TeslaManager

        var itemMultiPipe = HashMap<PipeTypeEnum, Item>()

        var teslaSupport: Boolean = false

        var RFSupport: Boolean = false

        var ic2Support: Boolean = false

        var rfPerEU: Double = 0.0

        lateinit var ic2Interface: IC2Interface

        lateinit var network: FRNetworkHandler
    }

}
